reconcile/utils/terrascript_client.py

```diff
@@ -3155,6 +3166,145 @@ def is_elasticsearch_domain_name_valid(name):
+                    continue
+                # res.get('', []) won't work, as publish_log_types is
+                # explicitly set to None if not set
+                log_types = res['publish_log_types']
+                if not log_types:
+                    log_types = []
+                for log_type in log_types:
+                    region = ns.get('cluster').get('spec').get('region')
+                    account = res['account']
+                    account_id = self.accounts[account]['uid']
+                    lg_identifier = \
+                        TerrascriptClient.elasticsearch_log_group_identifier(
+                            domain_identifier=res['identifier'],
+                            log_type=ElasticSearchLogGroupType(log_type),
+                        )
+                    tup = (account, region, account_id, lg_identifier)
+                    log_group_identifiers.append(tup)
+        return log_group_identifiers
+
+    def _get_tf_resource_elasticsearch_resource_policy(
+        self, account: str
+    ) -> aws_cloudwatch_log_resource_policy:
+        '''
+        https://docs.aws.amazon.com/opensearch-service/latest/developerguide/createdomain-configure-slow-logs.html
+        CloudWatch Logs supports 10 resource policies per Region.
+        If you plan to enable logs for several OpenSearch Service domains,
+        you should create and reuse a broader policy that includes multiple
+        log groups to avoid reaching this limit.
+        I.e., ideally we aggregate ALL log group identifiers for each
+        account first.
+        '''
+        log_group_identifiers = \
+            self._elasticsearch_aggregate_log_groups_per_account()
+
+        log_groups_policy = {
+            'Version': '2012-10-17',
+            'Statement': [{
+                'Effect': 'Allow',
+                'Principal': {
+                    'Service': 'es.amazonaws.com',
+                },
+                'Action': [
+                    'logs:PutLogEvents',
+                    'logs:CreateLogStream',
+                ],
+                'Resource': [
+                    f'arn:aws:logs:{tup[1]}:{tup[2]}:log-group:{tup[3]}:*'
+                    for tup in log_group_identifiers if tup[0] == account
+                ],
+            }]
+        }
+        log_groups_policy_values = {
+            'policy_name': 'es-log-publishing-permissions',
+            'policy_document': json.dumps(log_groups_policy, sort_keys=True),
+        }
+        resource_policy = aws_cloudwatch_log_resource_policy(
+            'es_log_publishing_resource_policy',
+            **log_groups_policy_values,
+        )
+        return resource_policy
+
+    def _get_tf_resource_elasticsearch_log_groups(
+        self, identifier: str, account: str,
+        resource: Mapping[str, Any], values: Mapping[str, Any],
+        output_prefix: str
+    ) -> Tuple[Iterable[Mapping[str, Any]], Iterable[Mapping[str, str]]]:
```

Comment:
As above, you want the types for your return values to be as precise as possible.

Reply:
Fixed